drivers/tee/optee/optee.c

```diff
@@ -177,9 +197,305 @@ static void u64_to_regs(uint64_t val, uint32_t *reg0, uint32_t *reg1)
+{
+	struct optee_notify *iter;
+
+	SYS_DLIST_FOR_EACH_CONTAINER(&data->notif, iter, node) {
+		if (iter->key == key) {
+			k_sem_give(&iter->wait);
+			return true;
+		}
+	}
+
+	return false;
+}
+
+static int optee_notif_wait(const struct device *dev, uint32_t key)
+{
+	int rc = 0;
+	struct optee_driver_data *data = dev->data;
+	struct optee_notify *entry;
+	k_spinlock_key_t sp_key;
+	int prev_val;
+
+	if (key > CONFIG_OPTEE_MAX_NOTIF)
+		return -EINVAL;
+
+	entry = k_malloc(sizeof(*entry));
+	if (!entry) {
+		return -ENOMEM;
+	}
+
+	k_sem_init(&entry->wait, 0, 1);
+	entry->key = key;
+
+	sp_key = k_spin_lock(&data->notif_lock);
+
+	/*
+	 * If notif bit was set then SEND command was already received.
+	 * Skipping wait.
+	 */
+	rc = sys_bitarray_test_and_clear_bit(&notif_bitmap, key, &prev_val);
+	if (rc || prev_val) {
+		goto out;
+	}
+
+	/*
+	 * If key is already registred, then skip.
+	 */
+	if (key_is_pending(data, key)) {
+		rc = -EBUSY;
+		goto out;
+	}
+
+	sys_dlist_append(&data->notif, &entry->node);
+
+	k_spin_unlock(&data->notif_lock, sp_key);
+	k_sem_take(&entry->wait, K_FOREVER);
+	sp_key = k_spin_lock(&data->notif_lock);
+
+	sys_dlist_remove(&entry->node);
+out:
+	k_spin_unlock(&data->notif_lock, sp_key);
+
+	k_free(entry);
+
+	return rc;
+}
+
+static void handle_cmd_notify(const struct device *dev, struct optee_msg_arg *arg)
+{
+	if (check_param_input(arg)) {
+		arg->ret = TEEC_ERROR_BAD_PARAMETERS;
+		return;
+	}
+
+	switch (arg->params[0].u.value.a) {
+	case OPTEE_RPC_NOTIFICATION_SEND:
+		if (optee_notif_send(dev, arg->params[0].u.value.b)) {
+			goto err;
+		}
+		break;
+	case OPTEE_RPC_NOTIFICATION_WAIT:
+		if (optee_notif_wait(dev, arg->params[0].u.value.b)) {
+			goto err;
+		}
+		break;
+	default:
+		goto err;
+	}
+
+	arg->ret = TEEC_SUCCESS;
+	return;
+
+err:
+	arg->ret = TEEC_ERROR_BAD_PARAMETERS;
+}
+
+static void handle_cmd_wait(const struct device *dev, struct optee_msg_arg *arg)
+{
+	if (check_param_input(arg)) {
+		arg->ret = TEEC_ERROR_BAD_PARAMETERS;
+		return;
+	}
+
+	k_sleep(K_MSEC(arg->params[0].u.value.a));
+
+	arg->ret = TEEC_SUCCESS;
+}
+
+static void free_shm_pages(void **pages)
+{
+	/*
+	 * Clean allocated pages if needed. Some function calls requires pages
+	 * allocation which should be freed after processing new request.
+	 * It is safe to free this list when another SHM op (e,g. another alloc
+	 * or free) was received.
+	 */
+	if (*pages) {
+		k_free(*pages);
+		*pages = NULL;
+	}
+}
+
+static uint32_t handle_func_rpc_call(const struct device *dev, struct tee_shm *shm,
+				     void **pages)
+{
+	struct optee_msg_arg *arg = shm->addr;
+
+	switch (arg->cmd) {
+	case OPTEE_RPC_CMD_SHM_ALLOC:
+		free_shm_pages(pages);
+		handle_cmd_alloc(dev, arg, pages);
+		break;
+	case OPTEE_RPC_CMD_SHM_FREE:
+		free_shm_pages(pages);
+		handle_cmd_free(dev, arg);
+		break;
+	case OPTEE_RPC_CMD_GET_TIME:
+		handle_cmd_get_time(dev, arg);
+		break;
+	case OPTEE_RPC_CMD_NOTIFICATION:
+		handle_cmd_notify(dev, arg);
```

Comment:
I hope this and below case isn't possible for atomic context

Reply:
Atomic context where exactly?

Reply:
in case if somebody invokes optee call under spin lock or from IRQ handler and the OS sends RPC
and request require some sleep

Reply:
It is programmer's error to invoke TEE from an atomic context. Exactly because TEE calls can be blocking.